code/modules/power/apc/apc_attack.dm

```diff
@@ -12,68 +14,100 @@
+	addtimer(CALLBACK(src, TYPE_PROC_REF(/atom, balloon_alert), user, "draining power"), ETHEREAL_APC_ALERT_DELAY)
+	while(do_after(user, ETHEREAL_APC_DRAIN_TIME, target = src))
+		if(isnull(used_stomach) || (used_stomach != user.get_organ_slot(ORGAN_SLOT_STOMACH)))
+			balloon_alert(user, "stomach removed!?")
 			return
-		if(stomach_cell.charge() > charge_limit)
-			addtimer(CALLBACK(src, TYPE_PROC_REF(/atom, balloon_alert), ethereal, "charge is full!"), alert_timer_duration)
+		if(isnull(cell))
+			balloon_alert(user, "cell removed!")
+			return
+		if(cell.charge() < half_max_charge)
+			balloon_alert(user, "safeties kicked in!")
+			return
+
+		var/our_available_charge = cell.charge() - half_max_charge
+		var/stomach_used_charge = stomach_cell.used_charge()
+		var/potential_charge = min(our_available_charge, stomach_used_charge)
+		var/to_drain = min(ETHEREAL_APC_POWER_GAIN, potential_charge)
+		var/energy_drained = cell.use(to_drain, force = TRUE)
+		used_stomach.adjust_charge(energy_drained)
+
+		if(stomach_cell.used_charge() <= 0)
+			balloon_alert(user, "your charge is full!")
+			return
+		if(cell.charge() <= 0)
+			to_chat(user, span_warning("APC is empty!"))
 			return
-		stomach.drain_time = world.time + ETHEREAL_APC_DRAIN_TIME
-		addtimer(CALLBACK(src, TYPE_PROC_REF(/atom, balloon_alert), ethereal, "draining power"), alert_timer_duration)
-		while(do_after(user, ETHEREAL_APC_DRAIN_TIME, target = src))
-			if(cell.charge <= (cell.maxcharge / 2) || (stomach_cell.charge() > charge_limit))
-				return
-			balloon_alert(ethereal, "received charge")
-			stomach.adjust_charge(ETHEREAL_APC_POWER_GAIN)
-			cell.use(ETHEREAL_APC_POWER_GAIN)
-		return
 
-	if(cell.charge >= cell.maxcharge - ETHEREAL_APC_POWER_GAIN)
-		addtimer(CALLBACK(src, TYPE_PROC_REF(/atom, balloon_alert), ethereal, "APC can't receive more power!"), alert_timer_duration)
+/// Handles charging our internal cell from an ethereal and their stomach
+/obj/machinery/power/apc/proc/charge_from_ethereal(mob/living/carbon/human/user, obj/item/organ/stomach/ethereal/used_stomach)
+	if(cell.charge() >= cell.max_charge())
+		addtimer(CALLBACK(src, TYPE_PROC_REF(/atom, balloon_alert), user, "APC full!"), ETHEREAL_APC_ALERT_DELAY)
 		return
-	if(stomach_cell.charge() < ETHEREAL_APC_POWER_GAIN)
-		addtimer(CALLBACK(src, TYPE_PROC_REF(/atom, balloon_alert), ethereal, "charge is too low!"), alert_timer_duration)
+	var/obj/item/stock_parts/power_store/stomach_cell = used_stomach.cell
+	if(stomach_cell.charge() <= 0)
+		addtimer(CALLBACK(src, TYPE_PROC_REF(/atom, balloon_alert), user, "charge is too low!"), ETHEREAL_APC_ALERT_DELAY)
 		return
-	stomach.drain_time = world.time + ETHEREAL_APC_DRAIN_TIME
-	addtimer(CALLBACK(src, TYPE_PROC_REF(/atom, balloon_alert), ethereal, "transfering power"), alert_timer_duration)
+
+	used_stomach.drain_time = world.time + ETHEREAL_APC_DRAIN_TIME
+	addtimer(CALLBACK(src, TYPE_PROC_REF(/atom, balloon_alert), user, "transferring power"), ETHEREAL_APC_ALERT_DELAY)
```

Comment:
```suggestion
	addtimer(CALLBACK(src, TYPE_PROC_REF(/atom, balloon_alert), user, "transferring power..."), ETHEREAL_APC_ALERT_DELAY)
```
Such a nitpick that I wouldn't even have posted it before merging if I didn't have another question.